Zend/zend_max_execution_timer.c

```diff
@@ -98,6 +109,16 @@ void zend_max_execution_timer_shutdown(void) /* {{{ */
 	if (err != 0) {
 		zend_strerror_noreturn(E_ERROR, errno, "Could not delete timer");
 	}
+
+	struct sigaction ign_sigrtmin;
+	ign_sigrtmin.sa_handler = SIG_IGN;
+	ign_sigrtmin.sa_flags = 0;
+	sigemptyset(&ign_sigrtmin.sa_mask);
+
+	if(sigaction(SIGRTMIN, &ign_sigrtmin, NULL) < 0) {
```

Comment:
Unfortunately this will reset the signal handler for the whole process.

Reply:
What do you mean by that? If the shutdown function is called, the process should be done with whatever PHP script it was working on. If the process wishes to start handling another PHP script, it will call `zend_max_execution_timer_init` again. Either way, the pending signals must be ignored to clear them, otherwise they will sit in SIGQ forever, even if another handler is set.

Reply:
Currently `zend_max_execution_timer_init()` and `zend_max_execution_timer_shutdown()` are called in `php_request_start()` and `php_request_shutdown()` respectively, so they are indeed called once per request. The signal handler is setup in `zend_set_timeout_ex()`, at least once per request.

However, when `zend_max_execution_timer_shutdown()` is called, there may be other requests executing in other threads of the same process. Resetting the signal handler here will affect them.

Reply:
Ah, I see that now. I don't think it is an issue though because in `zend_max_execution_timer_init()` we also reset the signal handler if shutdown has been previously called. Now that you've pointed this out, I think that was part of the reasoning for #12539, allowing timers to be reused, which would probably be out of scope here.

Reply:
The call to `sigaction(sev.sigev_signo, NULL, &EG(oldact));` in `zend_max_execution_timer_init()` is not resetting the handler, it's just copying the existing handler to `&EG(oldact)`. So by resetting handlers in `zend_max_execution_timer_shutdown()`, we are introducing a new issue.

We can not reset signals handlers at the end of each request, since this affects the whole process, including requests running in other threads. So I'm not sure how we can fix this issue if the only solution is to reset the signal handler.

However, I would like to challenge our understanding of the issue. Right now we are assuming that we leak signals when calling `timer_delete()` while some signals are pending. I would expect this to be a very rare occurrence, and it's unlikely that we are leaking 50k signals as shown in https://github.com/dunglas/frankenphp/issues/440, so I don't think we have found the actual cause of the issue.

`signal(7)` describes when signals are handled:

> Whenever  there  is a transition from kernel-mode to user-mode execution (e.g., on return from a system call or scheduling of a thread onto the CPU), the kernel checks whether there is a pending unblocked signal for which the process has established a signal handler.

For a signal to leak (assuming that `timer_delete()` causes the leak), we would need a signal to be queued before a `timer_delete()` call, and that no context switch happened between these two events, which is unlikely.

What seems more likely is that the signal is masked, which is why it may be queued.

An other possibility is that the signal queue is filled with nested queues created by `timer_create()`. `timer_create()` appears to [allocate a signal queue](https://github.com/torvalds/linux/blob/1f874787ed9a2d78ed59cb21d0d90ac0178eceb0/kernel/time/posix-timers.c#L407) upon invocation. This counts towards `RLIMIT_SIGPENDING`, and the function will return `EAGAIN` (Resource temporarily unavailable, the error you are getting) if `RLIMIT_SIGPENDING` is exceeded. SigQ in /proc/pid/status [reports RLIMIT_SIGPENDING usage](https://github.com/torvalds/linux/blob/master/fs/proc/array.c#L285) rather than the actual number of pending signals. So it's possible that the issue is only caused by the leak of one timer per thread.

Does [this change](https://github.com/php/php-src/blob/9c28adc73cc6997e81fd870879aeae38492f10ce/Zend/zend_max_execution_timer.c#L38-L42) _alone_ (without the signal handler reset) fixes the issue for you?

How many different threads had executed a PHP script when SigQ reached 50k?

Reply:
> However, I would like to challenge our understanding of the issue.

Good call here. I did, in-fact, fix the leak through `SIG_IGN` before discovering that we were calling `zend_max_execution_timer_init` twice. I actually didn't validate whether the `SIG_IGN` was still required after fixing that. 🤦 

> Does [this change](https://github.com/php/php-src/blob/9c28adc73cc6997e81fd870879aeae38492f10ce/Zend/zend_max_execution_timer.c#L38-L42) alone (without the signal handler reset) fixes the issue for you?

It does fix the issue. I've updated the PR.